### ISSUE TITLE
Flow unable to handle union of array types

### ISSUE THREAD

**Issue by Benjamin-Dobell** [facebook/flow#6773]
Flow getting itself into a strange state when encountering union of array types.

The errors are somewhat haywire, and aren't on the "problem line", but supposedly the type definitions. However, if you remove the problem line, the type definitions are just fine.

I'm running into this in a much more complicated real-world program, and it took me quite some time to track down the issue and then come up with a minimal reproduction case, but it is as follows:

```javascript
type AType = 'a';
type BType = 'b';
  
const A: AType[] = ['a'];
const B: BType[] = ['b'];

A.includes('a');
B.includes('b');

const aOrB = Math.random() ? A : B;
const test = aOrB.includes('a'); // Comment this out, and our type definitions are fine.
```

results in:

```
    5: const B: BType[] = ['b']
                ^ string literal `b` [1] is incompatible with string literal `a` [2].
        References:
        5: const B: BType[] = ['b']
                    ^ [1]
        4: const A: AType[] = ['a']
                    ^ [2]
```

[Try Flow](https://flow.org/try/#0C4TwDgpgBAggKuaBeKByAhqg3AKFJKAIQQJVQCNscoocBjAewDsBnYWALlhIgG0BdKCl4ZU-XI1btCXYogFCoIyuJw4YAOgCWTOgBsArgBMILABSiAlLkLbdhk+YqprayWyjoA8gCdCigFl0YAALDR90JiMGAFszSygAflgoWQlmD2BTdhRvPzt9Y1MLTGsoAHpyqABhWJiIJnZQrRYoBgNgABpPKLaDHyh8aBMAMx0tYC0Mzx9oMaYIDRwgA)

Note that the issue is specifically regarding the array union type `A[] | B[]` that's stored in `aOrB`.

If we replace our arrays with simple types but otherwise have a similar program structure, everything works just fine, see the [Try Flow](https://flow.org/try/#0C4TwDgpgBAggKuaBeKByAhqg3AKFJKAIQQJVQCNscoocBjAewDsBnYWALlhOTU10at2hLsURQylXDhgSkZfjkJzJVeszZR0AeQBOylAFl0wABYA6XeiYATBgFsAFAEooAflhRRAje2ARNFB19FT5sKAB6CKgAdQZdAGsWKAArAFdNADMASyYIHCA)

**Comment by wchargin** [comment]
Yep, member accesses on unions of generic classes have problems. (I’ve
seen this before, but can’t seem to find the references right now.)

Often, adding an explicit type annotation can fix the problem, with no
runtime cost. This works here:

```js
// @flow
type AType = 'a';
type BType = 'b';
  
const A: AType[] = ['a'];
const B: BType[] = ['b'];

A.includes('a');
B.includes('a');

const aOrB = Math.random() ? A : B;
const test = (aOrB.includes: (mixed) => boolean)('a');
```

[1]: https://flow.org/try/#0PTAEAEDMBsHsHcBQAXAngBwKagIIBUNsBeUAcgENSBuFQ0AIQK1BNICNrFRREBjWAHYBnZLgBcuJpgDaAXRahpFUrJr9ho+hMaE5CpR1WJEOAHQBLAb2gBXACaYhACmUBKGvQtXbD525p8giKg5ADyAE70CgCy5MgAFqbh5AJ2sAC2Tq6gAPy4oNpqQaLIjqIkTmGRXtb2jhJO6eYAHph22UQAfKBssLDQmCmuLpTuiEA

**Comment by wchargin** [comment]
> (I’ve seen this before, but can’t seem to find the references right now.)

Oh, here’s one: <https://github.com/facebook/flow/issues/6629>

My comments there about the covariance of `$ReadOnlyArray` may be helpful.

**Comment by gkz** [comment]
OP issue seems resolved
